genie-web/src/main/java/com/netflix/genie/web/configs/grpc/GenieGRpcServicesAutoConfiguration.java

```diff
@@ -0,0 +1,183 @@
+/*
+ *
+ *  Copyright 2017 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.configs.grpc;
+
+import com.netflix.genie.common.internal.configs.ProtoConvertersAutoConfiguration;
+import com.netflix.genie.common.internal.dto.v4.converters.JobDirectoryManifestProtoConverter;
+import com.netflix.genie.common.internal.dto.v4.converters.JobServiceProtoConverter;
+import com.netflix.genie.common.internal.util.GenieHostInfo;
+import com.netflix.genie.proto.FileStreamServiceGrpc;
+import com.netflix.genie.proto.HeartBeatServiceGrpc;
+import com.netflix.genie.proto.JobKillServiceGrpc;
+import com.netflix.genie.proto.JobServiceGrpc;
+import com.netflix.genie.proto.PingServiceGrpc;
+import com.netflix.genie.web.properties.GRpcServerProperties;
+import com.netflix.genie.web.rpc.grpc.interceptors.SimpleLoggingInterceptor;
+import com.netflix.genie.web.rpc.grpc.services.impl.v4.GRpcAgentFileStreamServiceImpl;
+import com.netflix.genie.web.rpc.grpc.services.impl.v4.GRpcHeartBeatServiceImpl;
+import com.netflix.genie.web.rpc.grpc.services.impl.v4.GRpcJobKillServiceImpl;
+import com.netflix.genie.web.rpc.grpc.services.impl.v4.GRpcJobServiceImpl;
+import com.netflix.genie.web.rpc.grpc.services.impl.v4.GRpcPingServiceImpl;
+import com.netflix.genie.web.rpc.grpc.services.impl.v4.JobServiceProtoErrorComposer;
+import com.netflix.genie.web.services.AgentJobService;
+import com.netflix.genie.web.services.AgentRoutingService;
+import com.netflix.genie.web.services.JobSearchService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.annotation.Order;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+
+/**
+ * Configures various gRPC services and related beans if gRPC functionality is enabled.
+ *
+ * @author tgianos
+ * @since 4.0.0
+ */
+@Configuration
+@ConditionalOnProperty(value = GRpcServerProperties.ENABLED_PROPERTY, havingValue = "true")
```

Comment:
This tangles these services creation to wether or not the server is enabled.
While it does make sense, it seems unnecessary, they don't actually depend on it.

Reply:
That is true and I'd like to remove it however it kind of messes up other things since we have certain entanglements that we should clean up. There are at least two service interfaces that are directly implemented by these service beans:

- `JobKillServiceV4` in `GRpcJobKillServiceImpl`
- `AgentFileStreamService` in `GRpcAgentFileStreamServiceImpl`

Implementations for these interfaces are required whether the gRPC server is started or not. So elsewhere in `GenieServicesAutoConfiguration` we have fallback anonymous implementations if gRPC is off.

This is kind of why I was asking if we can just leave gRPC on all the time as right now it's critical. If we disable this conditional then these beans would always be created and when those services are invoked and gRPC wasn't enabled it would behave really weird.

To me this smells like some tangling we should to fix to make the delineation between modules cleaner.  Let me know if that makes sense.